server/WAPI.rb

```diff
@@ -90,56 +83,89 @@ def self.base64_key_secret(key, secret)
     Base64.strict_encode64(ks)
   end
 
-  # use instance information to generate the proper url
+  ######### utilities for URL formatting
+  # use instance specific configuration information to generate the full url.
   def format_url(request)
     "#{@api_prefix}#{request}"
   end
 
-  def do_request(request)
+  # Responses may contain partial results.  In that case information about how to get the remaining data is returned
+  # in the 'Link' header. Link headers come back with explicit URLs pointing to Canvas servers.
+  # Remove that server added by the external service since we need to send queries back through the ESB rather than
+  # straight to the external service.
+  # This code assumes that the main portion of the next link (more_url) looks like the original request except that
+  # the beginning may be different (e.g. direct to canvas host vs through the ESB proxy) and the query parameters
+  # may be different (e.g. maybe add info about where to restart retrieval).
+
+  # A trivial more_url should become an empty string.
+  # Query parameters on the more_url should be passed through.
+  # Query parameters on the request_string should be ignored.
+  # The more_url is a complete URL that was returned from the external service.  The
+  # request_string it the partial url generated internally that doesn't have any explicit server / api
+  # version information.
+
+  def reduce_url(more_url, request_string)
+    logger.debug "#{self.class.to_s}:#{__method__}:#{__LINE__}: more_url: #{more_url} request_string: #{request_string}"
+
+    return "" if more_url.nil? || more_url.length == 0
+
+    # Get the main part of the original request without any query parameters.
+    main_request = request_string.index('?') ? request_string[/(^.+)\?/] : request_string
+
+    # Pull out the part of more_url that matches the original query and also pull along query parameters that were
+    # supplied in the more_url.
+    more_url = more_url[/#{main_request}.*/]
+
+    logger.debug "#{self.class.to_s}:#{__method__}:#{__LINE__}: reduced more_url: #{more_url}"
+    more_url
+  end
+
+  # Make a request to an external service and handle error conditions and headers.
+  def do_request(request_string)
 
     RestClient.log = logger if (logger.debug? and TRACE != FalseClass)
 
-    url=format_url(request)
+    # make the request specific to the separately configured API.
+    url=format_url(request_string)
```

Comment:
If the argument is a string (not an object as I expected `request` to be), what does it represent?  Is it the entire request as a single string or is it just the URL?  If it's just the URL, maybe name the variable "request_URL" or "request_URL_string".